p2p/stream/protocols/sync/protocol.go

```diff
@@ -0,0 +1,260 @@
+		chain:  config.Chain,
+		disc:   config.Discovery,
+		config: config,
+		ctx:    ctx,
+		cancel: cancel,
+		closeC: make(chan struct{}),
+	}
+	smConfig := streammanager.Config{
+		SoftLoCap: config.SmSoftLowCap,
+		HardLoCap: config.SmHardLowCap,
+		HiCap:     config.SmHiCap,
+		DiscBatch: config.DiscBatch,
+	}
+	sp.sm = streammanager.NewStreamManager(sp.ProtoID(), config.Host, config.Discovery,
+		sp.HandleStream, smConfig)
+
+	sp.rl = ratelimiter.NewRateLimiter(sp.sm, 50, 10)
+
+	sp.rm = requestmanager.NewRequestManager(sp.sm)
+
+	sp.logger = utils.Logger().With().Str("Protocol", string(sp.ProtoID())).Logger()
+	return sp
+}
+
+// Start starts the sync protocol
+func (p *Protocol) Start() {
+	p.sm.Start()
+	p.rm.Start()
+	p.rl.Start()
+	go p.advertiseLoop()
+}
+
+// Close close the protocol
+func (p *Protocol) Close() {
+	p.rl.Close()
+	p.rm.Close()
+	p.sm.Close()
+	p.cancel()
+	close(p.closeC)
+}
+
+// Specifier return the specifier for the protocol
+func (p *Protocol) Specifier() string {
+	return serviceSpecifier + "/" + strconv.Itoa(int(p.config.ShardID))
+}
+
+// ProtoID return the ProtoID of the sync protocol
+func (p *Protocol) ProtoID() sttypes.ProtoID {
+	return p.protoIDByVersion(MyVersion)
+}
+
+// Version returns the sync protocol version
+func (p *Protocol) Version() *version.Version {
+	return MyVersion
+}
+
+// Match checks the compatibility to the target protocol ID.
+func (p *Protocol) Match(targetID string) bool {
+	target, err := sttypes.ProtoIDToProtoSpec(sttypes.ProtoID(targetID))
+	if err != nil {
+		return false
+	}
+	if target.Service != serviceSpecifier {
+		return false
+	}
+	if target.NetworkType != p.config.Network {
+		return false
+	}
+	if target.ShardID != p.config.ShardID {
+		return false
+	}
+	if target.Version.LessThan(MinVersion) {
+		return false
+	}
+	return true
+}
+
+// HandleStream is the stream handle function being registered to libp2p.
+func (p *Protocol) HandleStream(raw libp2p_network.Stream) {
+	p.logger.Info().Str("stream", raw.ID()).Msg("handle new sync stream")
+	st := p.wrapStream(raw)
+	if err := p.sm.NewStream(st); err != nil {
+		// Possibly we have reach the hard limit of the stream
+		p.logger.Warn().Err(err).Str("stream ID", string(st.ID())).
+			Msg("failed to add new stream")
+		return
+	}
+	st.run()
+}
+
+func (p *Protocol) advertiseLoop() {
+	for {
+		sleep := p.advertise()
+		select {
+		case <-time.After(sleep):
+		case <-p.closeC:
+			return
+		}
+	}
+}
+
+// advertise will advertise all compatible protocol versions for helping nodes running low
+// version
+func (p *Protocol) advertise() time.Duration {
+	var nextWait time.Duration
+
+	pids := p.supportedProtoIDs()
+	for _, pid := range pids {
+		w, e := p.disc.Advertise(p.ctx, string(pid))
+		if e != nil {
+			p.logger.Warn().Err(e).Str("protocol", string(pid)).
+				Msg("cannot advertise sync protocol")
+			continue
+		}
+		if nextWait == 0 || nextWait > w {
+			nextWait = w
+		}
+	}
+	if nextWait < minAdvertiseInterval {
```

Comment:
advertise every minute?

Reply:
Nope. This is actually the backoff logic to handle error case.

If Advertise returns an error, the nextWait will return 0. If this is the case, the node will go into an infinite loop which is a black hole for CPU resource. So I added this minimum to retry in 1 min.

For normal non-error cases, nextWait will always be 6 hours from libp2p.